src/transform/src/semijoin_idempotence.rs

```diff
@@ -47,45 +69,108 @@ impl crate::Transform for SemijoinIdempotence {
         relation: &mut MirRelationExpr,
         _: TransformArgs,
     ) -> Result<(), crate::TransformError> {
-        {
-            // Iteratively descend the expression, at each node attempting to simplify a join.
-            let mut let_replacements = BTreeMap::<Id, Vec<Replacement>>::new();
-            let mut gets_behind_gets = BTreeMap::<Id, Vec<(Id, Vec<MirScalarExpr>)>>::new();
-            let mut worklist = vec![&mut *relation];
-            while let Some(expr) = worklist.pop() {
-                match expr {
-                    MirRelationExpr::Let { id, value, .. } => {
-                        let_replacements.insert(
-                            Id::Local(*id),
-                            list_replacements(&*value, &let_replacements, &gets_behind_gets),
-                        );
-                        gets_behind_gets
-                            .insert(Id::Local(*id), as_filtered_get(value, &gets_behind_gets));
+        let mut let_replacements = BTreeMap::<LocalId, Vec<Replacement>>::new();
+        let mut gets_behind_gets = BTreeMap::<LocalId, Vec<(Id, Vec<MirScalarExpr>)>>::new();
+        self.action(relation, &mut let_replacements, &mut gets_behind_gets)?;
+
+        mz_repr::explain::trace_plan(&*relation);
+        Ok(())
+    }
+}
+
+impl SemijoinIdempotence {
+    fn action(
```

Comment:
The `semijoin_idempotence.spec` test contains a minimal non-recursive example. Is it sufficient to demonstrate an entry in both `let_replacements` and `gets_behind_gets`? If so, can you copy-paste this as a docstring on top of the `action` method and state what will go into each context variable and why?

Reply:
Do you mean this file? https://github.com/MaterializeInc/materialize/pull/18714/files#diff-3c45034bbb3c4d9135b2de6b83806b98211e790a09b5519d0d9b1a8e1173ec14
The plan is not changing in that file, so SemijoinIdempotence is not kicking in there.

Reply:
I'll do a follow-up after we merge this.